File: java/Exercises.java

/**
 * Author: Wesley, Jody, Matthew, and Taylor
 * Collaborators: Ray Toal
 */

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.FileNotFoundException;
import java.lang.Math;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream().filter(predicate).findFirst().map(String::toLowerCase);
    }

    static record Sayer(String phrase) {

        Sayer and(String word) {
            // word = "" --> add space
            if (word.isEmpty()) {
                return new Sayer(this.phrase + " ");
            }
            // phrase starts with no words --> add 1st word
            if (this.phrase.isEmpty()) {
                return new Sayer(word);
            }
            // add a space + new word to the already made phrase
            return new Sayer(this.phrase + " " + word);
        }
    }

    public static Sayer say() {
        return new Sayer("");
    }

    public static Sayer say(String word) {
        return new Sayer(word);
    }

    static int meaningfulLineCount(String filename) throws IOException {
        try (FileReader in = new FileReader(filename)) {
            BufferedReader br = new BufferedReader(in);
            return (int) br.lines()
                           .filter(line -> !line.trim().isEmpty())
                           .filter(line -> line.trim().charAt(0) != '#')
                           .count();
        } catch (IOException e) {
            throw new FileNotFoundException("No such file");
        }
    }
}

record Quaternion(double a, double b, double c, double d) {

    static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    static final Quaternion I = new Quaternion(0, 1, 0, 0);
    static final Quaternion J = new Quaternion(0, 0, 1, 0);
    static final Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    public Quaternion plus(Quaternion other) {
        return new Quaternion(
                a + other.a,
                b + other.b,
                c + other.c,
                d + other.d);
    }

    public Quaternion times(Quaternion other) {
        return new Quaternion(
                a * other.a - b * other.b - c * other.c - d * other.d,
                a * other.b + b * other.a + c * other.d - d * other.c,
                a * other.c - b * other.d + c * other.a + d * other.b,
                a * other.d + b * other.c - c * other.b + d * other.a);
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();

        if (a != 0) {
            sb.append(a);
        }

        double[] coeffs = { b, c, d };
        String[] symbols = { "i", "j", "k" };
        for (int i = 0; i < coeffs.length; i++) {
            double coeff = coeffs[i];
            String symbol = symbols[i];
            if (coeff != 0) {
                if (coeff > 0 && sb.length() > 0) {
                    sb.append("+");
                }
                if (Math.abs(coeff) == 1) {
                    sb.append(coeff == -1 ? "-" + symbol : symbol);
                } else {
                    sb.append(coeff + symbol);
                }
            }
        }

        if (sb.length() == 0) {
            return "0";
        }
        return sb.toString();
    }

}

sealed interface BinarySearchTree permits Empty, Node {
    int size();

    boolean contains(String value);

    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) > 0) {
            return new Node(this.value, left, right.insert(value));
        } else {
            return new Node(this.value, left.insert(value), right);
        }
    }

    @Override
    public String toString() {
        String leftStr = (left instanceof Empty) ? "" : left.toString();
        String rightStr = (right instanceof Empty) ? "" : right.toString();
        return "(" + leftStr + value + rightStr + ")";
    }
}

File: kotlin/Exercises.kt
/**
 * Author: Wesley, Jody, Matthew, and Taylor
 * Collaborators: Ray Toal
 */

import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import kotlin.math.abs

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

fun firstThenLowerCase(words: List<String>, predicate: (it: String) -> Boolean): String? {
    /**
     *  Finds the first word in the list that follows the given predicate,
     *  then returns that word lowercased.
     */
    return words.filter(predicate).getOrNull(0)?.lowercase()
}

data class PhraseConstructor(val phrase: String) {
    fun and(excerpt: String): PhraseConstructor {
        return PhraseConstructor(phrase + " " + excerpt)
    }   
}

fun say(phrase: String = "") : PhraseConstructor {
    /**
     *  A compounding method starting with say functions, then multiple
     *  and functions and ends with calling phrase. Creates a sentence where
     *  each phrase is added to the sentence with spaces in between.
     */
    return PhraseConstructor(phrase)
}

@Throws(IOException::class)
fun meaningfulLineCount(filePath: String): Long {
    /**
     *  Counts all of the lines in a given file where it contains
     *  characters (not whitespaces) and does not begin with a hashtag
     *  symbol (#).
     */
    var counter: Long = 0L
    BufferedReader(FileReader(filePath)).use {
        input ->
        var line: String? 
        while (input.readLine().also {line = it} != null) {
            val stripped: String = line!!.replace("\\s".toRegex(),"")
            if (stripped.length > 0 && stripped[0] != '#') {
                ++counter
            }
        }
    }
    return counter
}

data class Quaternion(
    val a: Double, 
    val b: Double, 
    val c: Double, 
    val d: Double
) {
    init {
        require(!a.isNaN() && !b.isNaN() && !c.isNaN() && !d.isNaN()){
            "Numbers can not be NaN"
        }
    }
    
    fun coefficients() : List<Double> {
        return listOf(a, b, c, d)
    }

    fun conjugate() : Quaternion {
        return Quaternion(a, -b, -c, -d)
    }

    operator fun plus(other: Quaternion) : Quaternion {
        return Quaternion(
            a + other.a,
            b + other.b,
            c + other.c,
            d + other.d
        )
    }

    operator fun times(other: Quaternion) : Quaternion {
        return Quaternion(
            a * other.a - b * other.b - c * other.c - d * other.d,
            a * other.b + b * other.a + c * other.d - d * other.c,
            a * other.c + c * other.a - b * other.d + d * other.b,
            a * other.d + d * other.a + b * other.c - c * other.b
        )
    }

    override fun equals(other: Any?) : Boolean {
        if (this === other) return true
        if (other !is Quaternion) return false
        return a == other.a && b == other.b && c == other.c && d == other.d
    }

    override fun toString() : String {
        // Base case: 
        if (this == Quaternion.ZERO) {
            return "0"
        }

        // Creates a (almost) perfect Quaternion string!
        val constant: String = if (a != 0.0) "${a}" else ""
        val equation: String = constant + innerNominal(b, "i") + innerNominal(c, "j") + innerNominal(d, "k")

        // If the equation begins with a "+", remove it before returning
        return if (equation.firstOrNull() == '+') equation.substring(1) else equation
    }

    private fun innerNominal(num: Double, variable: String) : String{
        val sign: String = if (num > 0.0) "+" else (if (num < 0.0) "-" else "")
        val nominal: String = if (num == 0.0) "" else (if (Math.abs(num) == 1.0) "$variable" else "${Math.abs(num)}$variable")
        return sign + nominal
    }
    
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
}

sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(target: String): Boolean
    fun insert(itemValue: String) : BinarySearchTree
    override fun toString(): String

    data class TreeNode(
        val itemValue: String, 
        val left: BinarySearchTree = BinarySearchTree.Empty, 
        val right: BinarySearchTree = BinarySearchTree.Empty
    ) : BinarySearchTree {

        override fun size() : Int {
            return left.size() + 1 + right.size()  
        }

        override fun contains(target: String) : Boolean {
            return itemValue == target || left.contains(target) || right.contains(target)
        }

        override fun insert(newitemValue: String) : BinarySearchTree { 
            val comparison = newitemValue.compareTo(itemValue)
            if (comparison < 0) {
                return TreeNode(this.itemValue, left.insert(newitemValue), right)
            } else if (comparison > 0) {
                return TreeNode(this.itemValue, left, right.insert(newitemValue))
            } else {
                return TreeNode(this.itemValue, left, right)
            }
        }

        override fun toString(): String {
            val leftString = if (left != BinarySearchTree.Empty) "${left}" else "" 
            val rightString = if (right != BinarySearchTree.Empty) "${right}" else "" 
            return "(${leftString}${itemValue}${rightString})"
        }
    }

    object Empty : BinarySearchTree {
        override fun size() : Int {
            return 0
        }
        
        override fun contains(target: String) : Boolean {
            return false
        }

        override fun insert(itemValue: String) : BinarySearchTree {
            val root: BinarySearchTree = TreeNode(itemValue)
            return root
        }

        override fun toString() : String {
            return "()"
        }
    }

}